worker/include/RTC/RateCalculator.hpp

```diff
@@ -14,10 +14,14 @@ namespace RTC
 	public:
 		static constexpr size_t DefaultWindowSize{ 1000u };
 		static constexpr float DefaultBpsScale{ 8000.0f };
+		static constexpr uint16_t DefaultWindowItems{ 1000u };
 
 	public:
-		RateCalculator(size_t windowSize = DefaultWindowSize, float scale = DefaultBpsScale)
-		  : windowSize(windowSize), scale(scale)
+		RateCalculator(
+		  size_t windowSize    = DefaultWindowSize,
+		  float scale          = DefaultBpsScale,
+		  uint16_t windowItems = DefaultWindowItems)
```

Comment:
Is there any reason/case for `windowItems` value to be different than `windowSize` value?

I see no setter for it, and it takes the same value as `windowSize` in constructor, by default.

Reply:
I wonder the very same.

Reply:
The `windowItems` value depends on both the rate evaluation window period (`windowSize`) and the events frequency. 
For instance, if the measured event frequency is 1 every 2ms, you can set `windowSize=1000` windowItems=500`.

Reply:
`windowSize` is only used to get the `scale` factor and to retrieve the new `oldestTimeIndex` whereas `windowItems` is the range (from to `windowItems`) where the items are really stored in the `buffer`.

Having a `windowSize` bigger than the `windowItems` does not provide accurate results I would say. Ie:

````
windowSize = 1000
|-----------------------------------------------------|

windowItems = 500
|------------------------|

````

We are calculating the rate for a time window of 1 second, and only considering the data of the last 1/2 second, which is not accurate.

Due to that rationale I think `windowItems` should be removed. Any comment on this @vpalmisano?

Reply:
The `windowSize` is expressed in seconds, while the `windowItems` is the maximum number of items that can be stored at the same time in the time window, so in your example you must take into account also the frequency of the data sample. 
If you receive 1 sample per millisecond, you will fill the array with 500 items considering only 1/2 second of data. 
Instead, if the data sample frequency is 1 every 2 milliseconds, a 500 items array will correspond to 1 second of data.
I can perform same tests evaluating if the window is filled or not when changing the `windowItems` value.

Reply:
I've added a warning to detect when the calculation buffer is filled.
With the default `windowSize = 2500` used in `RtpDataCounter` with `windowItems >= 700` I get no messages. I think that `windowItems = 1000` is a safe value that we can keep.

Reply:
Thanks @vpalmisano,

I see. Usually we don't control the frequency at which we update the rate calculator, that's why there was 1 item representing each millisecond.

Reply:
Yes, using the log warning we will able to see if the buffer size is underestimated.

Reply:
In the real world test I'm getting tons of logs saying,
 `mediasoup:WARN:Channel [pid:92750] RTC::RateCalculator::Update() | calculation buffer full, windowSize:2500 ms windowItems:1000 +0ms`

Reply:
I assume this is not good news. Is there something than can be fixed/improved here?

Reply:
I suppose the old approach with fixed side buckets is much reliable. It can be tuned a bit to help compiler to vectorise calculation sum of removed buckets and memset. 

There is a simple math here:
2 Mbit/s video stream gives on average 200 packets per sec, when keyframe is generated, it can be up to 300. 5 participants room can generate 1200 packets per sec for the rate calculator if transport-cc is enabled. That can already overfill the calculation buffer, especially for window size 2.5 sec. And that is us just for 5 participants.

The old fixed size bucket implementation is resistant to this, if window size of 2.5 sec is need, it can be achieved by using 2.5 ms buckets without increasing the buffer size.

Reply:
This means that with more than 10 participants and/or more than 2mbps also with the old approach we were underestimating the calculated rates. We need to find a different algorithm that allows to increase the number of packets per second without increasing the bucket size.

Reply:
On the meantime we can set the bucket size to 2500, to obtain at least the same performance of the old approach.

Reply:
Old algorithm has no issues with multiple packets with the same ms timestamp, those packets would go to the same bucket, and bucket would contain sum of all those packets sizes. Buckets are for the timeslots, not for the packets.

New algorithm with bucket size 2500 would have worse performance, then the old one, it tracks timestamp together with packet's sizes and less SIMD friendly.

Reply:
We can modify the new algorithm to sum more packets with timestamps that not differ from a minimum delta to the same array position, keeping the array size to a small value.

Reply:
Same can be applied to the old algorithm. Bucket duration can be any integer. For example 2.5 second window can be handled by 500 buckets with 5ms duration or with 250 buckets with 10ms duration.

Reply:
Thanks for the info, guys. So AFAIU the problem has always been there (even before this PR): if bitrate is very high it may happen that there is no buffer enough to store N seconds of data, is that right? And the only "problem" after this PR is that the size of such a buffer is now smaller so making it bigger would resolve the problem. Did I understand it properly?

Reply:
I've added some modifications here (https://github.com/versatica/mediasoup/compare/v3...vpalmisano:optimize-rate-calculator) to sum more packets with timestamps that not differ from a minimum delta to the same array position.
I'm testing it with a large (>100) number of consumers, I will post the results very soon.

Reply:
> We have to decide the calculation sliding window granularity. E.g. with a 2500ms window we can use a 250 items array, obtaining a granularity of 10ms. I will push a commit very soon.

Right, based on the window size (ms) and number of items in the window we must compute how much time each item represents. :+1:

Reply:
> So AFAIU the problem has always been there (even before this PR):

No, we didn't have that problem before because each item represented a millisecond and each item would hold the sum of the sizes represented.

Reply:
@jmillan right, the problem was not there before this PR. The drawback of the old approach is that it is using a number of items == window size in ms and in some cases this size could be equal to 2500 or 6000 packets.

Can we re-open this PR or should I open a new one with the latest fixes?

Reply:
PRs cannot be reopened AFAIK so create a new one, please.

Reply:
@notorca thanks for reporting the bug and comments.
Fixed now in: https://github.com/versatica/mediasoup/pull/547

Reply:
Nice, I will test it soon.

Reply:
@vpalmisano Works good for me.